### DISCUSSION TITLE
How to extend selection with search?

### DISCUSSION THREAD

**Discussion by wellle** [helix-editor/helix#5668] [General] [answered]
[I posted this originally on Reddit](https://www.reddit.com/r/HelixEditor/comments/10cty8p/how_to_extend_selection_with_search)

There is one major use case I really struggle with in Helix: As an example consider a long file with multiple section markers like this:
```
... many lines ...

// section 4.7

... many lines ...

// section 4.8

... many lines ...
```

Let's say I just want to delete all of section `4.7`. In Vim I could do something like `/4.7<ret>v/4.8<ret>d`. That is in Vim I can use search to move one end of my selection to the next search match.
When I try to do the same in Helix, searching in Select mode seems to reset my current selection and always replaces it with the next search match.
How do I select from one search match to another search match? So in my example from `4.7` to `4.8`? Thank you!

---

Side note: Why does `/` behave in this way in the first place? I've had other use cases where I had multiple selections and wanted to advance all selections to a certain word in the line, so instinctively I tried to search for that word expecting each selection to jump forward (or extend if in selection mode) to the next respective match. This seems very useful to me, but is broken by how `/` dismisses all current selections. For a powerful multi selection Editor I think Helix would benefit from handling searching with multiple selections.

Thank you!

**Comment by the-mikedavis** [comment, marked as answer]
This has been discussed in the Matrix chat here and there: it would be useful to have another form of search that behaves like `t` or `f` do. I posted https://github.com/helix-editor/helix/issues/5672 with some design/implementation context to track that. The current select-mode behavior for search can be helpful to create multiple selections so I think this should be a new feature that doesn't touch the existing search.

Say that the binding for that was `A-/`: you could search for the "4.7" with `/4.7<ret>`, then extend the search with `v<A-/>4.8<ret>` and then extend the whole selection to line bounds with `X`.

About the subreddit: the maintainers aren't affiliated with it. The [Matrix channel](https://github.com/helix-editor/helix#getting-help) and GitHub repo are the best way to get help from the community.

**Reply by wellle** [reply]
[Sorry, first time using GitHub discussions, moving my responses from below to here]

Thank you, this is very helpful and seems like a good approach 👍

**Reply by wellle** [reply]
@the-mikedavis: Another related use case I just ran into:
1. I selected some lines I wanted to operate on.
2. I split the selection by line.
3. I did some modifications to each line (with one selection per line)
4. I press <kbd>x</kbd> to extend all selections to full lines again.
5. Now I want to `:reflow` the lines.

This doesn't work as its reflowing each line (each selection) individually. The missing step between 4. and 5. would be to join/merge all my selections back into a single large selection.

(In this specific case I could have saved my selections before splitting them with `<c-s>` and then go back with `<c-o>`, but I didn't have the foresight)

How is that related to this discussion? If we had a way of joining/merging selections by selecting from the start of the first selection all the way up to the end of the last selection, this would also add a way of what I was asking for in the original message above: I could create two selections, one on `4.7` and one on `4.8`:

```
/4.7<ret>C/4.8<ret>
```
[`C` is used to create second selection which is then affected by the second search]

and then invoke the `join/merge-selections` command (via `<space>?` might be enough) to end up with a single selection from `4.7` to `4.8`. 

What do you think? Is there already a way to do this? Could this be a useful addition? Thank you!

**Reply by the-mikedavis** [reply]
For that you can use `A-_` which merges any consecutive selections. It might be useful to create one selection that spans from the beginning of the first selection to the end of the last selection allowing gaps in between. That is implemented in #3720 as `^j`

**Reply by wellle** [reply]
That sounds great! I just tried `<alt-_>` but it doesn't seem to do anything, also on https://docs.helix-editor.com/keymap.html#select--extend-mode it doesn't seem to be documented yet. I'm on version `22.12`. Would this be only available on builds from master? 

If it's not too much of a bother, could you show me the PR which added `<alt-_>`? Thanks again!
Edit: Found it: https://github.com/helix-editor/helix/pull/5047, merged a day after the release I'm using 🙂, I'll try building from master soon.

**Reply by wellle** [reply]
Okay, I installed the latest master version and have `<alt-a>` now. It works for the example in the thread, but not for the original case. But it looks like `^j` would be a fit for that if I understand the conversation over there correctly.

**Comment by rcorre** [comment]
This isn't a perfect solution, but I've found the following useful:

```toml
[keys.select]
n = ["extend_search_next", "merge_selections"]
N = ["extend_search_prev", "merge_selections"]
```

That makes search work the way I'd expect in select mode as long as I only have one cursor. It can backfire if you have multiple cursors before you start the search, so a solution to #5672 would still be ideal.

**Comment by canadaduane** [comment]
Newcomer here. I expected `v/` (followed by a search term) to extend the selection with search. Would this be a reasonable expectation?

Example:

```
|hello
text
world

more text
```

`v/world` would highlight from current cursor (before `hello`) to end of `world`.

**Reply by kaos** [reply]
I had the same expectation, but found that the search hits creates new cursors, which also makes sense.

An alternative approach to achieve the above, is if jumping and not only movement would extend the selection in selection mode (as long as the jump stays in the same buffer). Then this would work for the above example: `c-s/world<ret>vc-o` (Save current position in jumplist, search for `world`, enter selection mode, jump to previous jumplist entry.)
